### PR TITLE
Allow disabling optimization that doesn't report timers below threshold

### PR DESCRIPTION
**Pull request by vilhena-stripe**
### Motivation
Allow us to write tests involving timers without having to use `sleep`s so timers are reported

### CODE REVIEW

File: common/Timer.cc [hunk 1]
```diff
@@ -44,6 +44,8 @@ microseconds Timer::clock_gettime_coarse() {
 Timer::Timer(spdlog::logger &log, ConstExprStr name, FlowId prev, initializer_list<pair<ConstExprStr, string>> args,
              microseconds start, initializer_list<int> histogramBuckets)
     : log(log), name(name), prev(prev), self{0}, start(start) {
+    keepShortTimers = false;
```
Comment (DarkDimius) [inline common/Timer.cc:47]: Let's not modify static field in instance methods

Comment (DarkDimius) [reply common/Timer.cc:47]: It should probably have a separate initializer in this file.

File: common/Timer.h [hunk 2]
```diff
@@ -58,6 +58,9 @@ class Timer {
 
     static microseconds clock_gettime_coarse();
 
+    // Disable dropping timers with duration below clock_threshold_coarse for tests
+    void disableDroppingShortTimersForTests();
```
Comment (DarkDimius) [inline common/Timer.h:62]: Let's land usage in the same PR?
Will make it easier to discuss motivation and thus if this is the right approach.

Comment (DarkDimius) [reply common/Timer.h:62]: (Changing such global configs in tests feels problematic for me, but I lack context to know what are we looking to address)

Comment (jvilk-stripe) [reply common/Timer.h:62]: Context: We have tests that ensure that LSP is reporting timing metrics correctly.

However, timing metrics are dropped if they are below a certain threshold.

I've inserted sleeps into some tests to ensure that expected timers are reported, but it's very brittle.

This mechanism is intended to let us _disable_ the timer optimization so we can remove those sleeps and make these few metrics tests robust and fast.

File: common/Timer.cc (hunk 1 above)
Comment (jvilk-stripe) [reply common/Timer.cc:47]: ==, this should be a `private` `static` property on the `Timer` class.

File: common/Timer.cc [hunk 3]
```diff
@@ -141,10 +143,14 @@ void Timer::setTag(ConstExprStr name, ConstExprStr value) {
     tags->push_back(make_pair(name, value));
 }
 
+void Timer::disableDroppingShortTimersForTests() {
```
Comment (jvilk-stripe) [inline common/Timer.cc:146]: This should be a `public` `static` method

File: common/Timer.h [hunk 4]
```diff
@@ -74,6 +77,7 @@ class Timer {
     // on a bucket.
     std::unique_ptr<std::vector<int>> histogramBuckets;
     bool canceled = false;
+    static bool keepShortTimers;
```
Comment (jvilk-stripe) [inline common/Timer.h:80]: This should be explicitly initialized to `false`.

File: test/lsp/multithreaded_protocol_test_corpus.cc [hunk 5]
```diff
@@ -618,8 +615,8 @@ TEST_CASE_FIXTURE(MultithreadedProtocolTest, "CanceledRequestsDontReportLatencyM
     sendAsync(LSPMessage(make_unique<NotificationMessage>("2.0", LSPMethod::PAUSE, nullopt)));
     // Send a hover to request the documentation string.
     sendAsync(*hover("foo.rb", 2, 6));
-    // Pause so that all latency timer for the hover is >1ms and would normally get reported.
-    this_thread::sleep_for(chrono::milliseconds(2));
+    // Ensure all latency timers are reported.
+    Timer::disableDroppingShortTimersForTests();
```
Comment (jez) [inline test/lsp/multithreaded_protocol_test_corpus.cc:619]: We're calling this in four places but never cleaning up after ourselves. Are we introducing flakiness into these tests?

I think we should either set this once at the start of the test and have it apply to everything, or have something that automatically cleans up the state when a variable goes out of scope.

Also I'm not super familiar with this test, but it says `multithreaded` in the name. This threshold is not stored in a threadlocal and is not guarded by synchronization code, but we're mutating it multiple times. Maybe that's fine, maybe it's a problem—jvilk would know more.

Comment (jvilk-stripe) [reply test/lsp/multithreaded_protocol_test_corpus.cc:619]: > Also I'm not super familiar with this test, but it says multithreaded in the name. This threshold is not stored in a threadlocal and is not guarded by synchronization code, but we're mutating it multiple times. Maybe that's fine, maybe it's a problem—jvilk would know more.

Good point; I should've caught this. We should be calling this once at the start of the test suites that involve timers.

**Comment by vilhena-stripe** [comment]
Closing this PR in favor of a solution where we report all latency timers instead

**Pull request closed without merging**